pkg/fs/removeall_windows.go

```diff
@@ -0,0 +1,214 @@
+				// We don't want to recurse into a name surrogate (e.g. symlink) because we will end up deleting stuff elsewhere on the system.
+				// In this case there's nothing else to do for this entry, so just return an error.
+				return pathErr("RemoveDirectory", path, err)
+			}
+			// Wrapping in an anonymous function so that the deferred FindClose call is scoped properly.
+			// Iterate the directory and remove all children by recursing into removeAll.
+			if err := func() error {
+				var fd findData
+				// findFirstFileExW allows us to avoid the shortname lookup. See comment at top of file for details.
+				pattern := join(path, []uint16{'*'})
+				find, err := findFirstFileExW(terminate(pattern), findExInfoBasic, &fd, findExSearchNameMatch, nil, 0)
+				if err == windows.ERROR_FILE_NOT_FOUND {
+					// No children is weird, because we should always have "." and "..".
+					// If we do hit this, just continue to the next deletion attempt.
+					return nil
+				} else if err != nil {
+					return pathErr("FindFirstFileEx", pattern, err)
+				}
+				defer windows.FindClose(find)
+				for {
+					var child []uint16
+					child, err = truncAtNull(fd.name[:])
+					if err != nil {
+						return err
+					}
+					if !equal(child, []uint16{'.'}) && !equal(child, []uint16{'.', '.'}) { // Ignore "." and ".."
+						if err := removeAll(join(path, child), fd.attributes, fd.reserved0); err != nil {
+							return err
+						}
+					}
+					err = findNextFileW(find, &fd)
+					if err == windows.ERROR_NO_MORE_FILES {
+						break
+					} else if err != nil {
+						return pathErr("FindNextFile", path, err)
+					}
+				}
+				return nil
+			}(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func getFileInfo(path []uint16) (attrs uint32, reparseTag uint32, _ error) {
+	h, err := windows.CreateFile(terminate(path), 0, 0, nil, windows.OPEN_EXISTING, windows.FILE_OPEN_REPARSE_POINT|windows.FILE_FLAG_BACKUP_SEMANTICS, 0)
+	if err != nil {
+		return 0, 0, pathErr("CreateFile", path, err)
+	}
+	defer windows.CloseHandle(h) //nolint:errcheck
+	var ti fileAttributeTagInfo
+	if err = windows.GetFileInformationByHandleEx(h, windows.FileAttributeTagInfo, (*byte)(unsafe.Pointer(&ti)), uint32(unsafe.Sizeof(ti))); err != nil {
+		return 0, 0, pathErr("GetFileInformationByHandleEx", path, err)
+	}
+	return ti.attributes, ti.tag, nil
+}
+
+func equal(v1, v2 []uint16) bool {
+	if len(v1) != len(v2) {
+		return false
+	}
+	for i := range v1 {
+		if v1[i] != v2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func join(parent, child []uint16) []uint16 {
+	return append(append(parent, '\\'), child...)
+}
+
+func pathErr(op string, path []uint16, err error) error {
+	return &fs.PathError{Op: op, Path: windows.UTF16ToString(path), Err: err}
+}
+
+// terminate takes a []uint16 and returns a null-terminated *uint16.
+func terminate(path []uint16) *uint16 {
+	return &append(path, '\u0000')[0]
```

Comment:
should we check that this path isnt already null-terminated?

Reply:
I don't think there's any valid case where we would pass a null terminated string here, since the idea is that we only terminate right as we pass it to a syscall. Maybe we could add a check, but it would be the equivalent of a debug assert semantically.